File: python/mops/src/mops/opsa.py
import numpy as np

from . import _dispatch
from .checks import _check_opsa, _check_opsa_vjp, _check_opsa_vjp_vjp
from .dispatch_operation import dispatch_operation
from .utils import mops_tensor, null_mops_tensor_like


def outer_product_scatter_add(A, B, indices_output, output_size):
    _check_opsa(A, B, indices_output, output_size)

    A = _dispatch.make_contiguous(A)
    B = _dispatch.make_contiguous(B)
    indices_output = _dispatch.make_contiguous(indices_output)
    indices_output = indices_output.astype(np.int32)

    output = _dispatch.empty_like((output_size, A.shape[1], B.shape[1]), A)

    function = dispatch_operation(
        "outer_product_scatter_add",
        A,
    )

    function(
        mops_tensor(output),
        mops_tensor(A),
        mops_tensor(B),
        mops_tensor(indices_output),
    )

    return output


def outer_product_scatter_add_vjp(
    grad_output,
    A,
    B,
    indices_output,
    compute_grad_A=True,
    compute_grad_B=True,
):
    _check_opsa_vjp(grad_output, A, B, indices_output, grad_output.shape[0])

    grad_output = _dispatch.make_contiguous(grad_output)
    A = _dispatch.make_contiguous(A)
    B = _dispatch.make_contiguous(B)
    indices_output = _dispatch.make_contiguous(indices_output)
    indices_output = indices_output.astype(np.int32)

    if compute_grad_A:
        grad_A = _dispatch.empty_like(A.shape, A)
        mops_grad_A = mops_tensor(grad_A)
    else:
        grad_A = None
        mops_grad_A = null_mops_tensor_like(A)

    if compute_grad_B:
        grad_B = _dispatch.empty_like(B.shape, B)
        mops_grad_B = mops_tensor(grad_B)
    else:
        grad_B = None
        mops_grad_B = null_mops_tensor_like(B)

    function = dispatch_operation(
        "outer_product_scatter_add_vjp",
        A,
    )

    function(
        mops_grad_A,
        mops_grad_B,
        mops_tensor(grad_output),
        mops_tensor(A),
        mops_tensor(B),
        mops_tensor(indices_output),
    )

    return grad_A, grad_B


def outer_product_scatter_add_vjp_vjp(
    grad_grad_A,
    grad_grad_B,
    grad_output,
    A,
    B,
    indices_output,
    compute_grad_grad_output=True,
    compute_grad_A_2=True,
    compute_grad_B_2=True,
):
    _check_opsa_vjp_vjp(
        grad_grad_A,
        grad_grad_B,
        grad_output,
        A,
        B,
        indices_output,
        grad_output.shape[0],
    )

    grad_grad_A = _dispatch.make_contiguous(grad_grad_A)
    grad_grad_B = _dispatch.make_contiguous(grad_grad_B)
    grad_output = _dispatch.make_contiguous(grad_output)
    A = _dispatch.make_contiguous(A)
    B = _dispatch.make_contiguous(B)
    indices_output = _dispatch.make_contiguous(indices_output)
    indices_output = indices_output.astype(np.int32)

    if compute_grad_grad_output:
        grad_grad_output = _dispatch.empty_like(
            (grad_output.shape[0], A.shape[1], B.shape[1]), grad_output
        )
        mops_grad_grad_output = mops_tensor(grad_grad_output)
    else:
        grad_grad_output = None
        mops_grad_grad_output = null_mops_tensor_like(grad_output)

    if compute_grad_A_2:
        grad_A_2 = _dispatch.empty_like(A.shape, A)
        mops_grad_A_2 = mops_tensor(grad_A_2)
    else:
        grad_A_2 = None
        mops_grad_A_2 = null_mops_tensor_like(A)

    if compute_grad_B_2:
        grad_B_2 = _dispatch.empty_like(B.shape, B)
        mops_grad_B_2 = mops_tensor(grad_B_2)
    else:
        grad_B_2 = None
        mops_grad_B_2 = null_mops_tensor_like(B)

    function = dispatch_operation(
        "outer_product_scatter_add_vjp_vjp",
        A,
    )

    function(
        mops_grad_grad_output,
        mops_grad_A_2,
        mops_grad_B_2,
        (
            mops_tensor(grad_grad_A)
            if grad_grad_A is not None
            else null_mops_tensor_like(A)
        ),
        (
            mops_tensor(grad_grad_B)
            if grad_grad_B is not None
            else null_mops_tensor_like(B)
        ),
        mops_tensor(grad_output),
        mops_tensor(A),
        mops_tensor(B),
        mops_tensor(indices_output),
    )

    return grad_grad_output, grad_A_2, grad_B_2

File: mops-torch/src/opsa.cpp
#include "mops/torch/opsa.hpp"
#include "mops/torch/utils.hpp"

using namespace mops_torch;

torch::Tensor mops_torch::outer_product_scatter_add(
    torch::Tensor A, torch::Tensor B, torch::Tensor indices_output, int64_t output_size
) {
    return OuterProductScatterAdd::apply(A, B, indices_output, output_size);
}

torch::Tensor OuterProductScatterAdd::forward(
    torch::autograd::AutogradContext* ctx,
    torch::Tensor A,
    torch::Tensor B,
    torch::Tensor indices_output,
    int64_t output_size
) {
    details::check_all_same_device({A, B, indices_output});
    details::check_floating_dtype({A, B});
    details::check_integer_dtype({indices_output});
    details::check_number_of_dimensions(A, 2, "A", "outer_product_scatter_add");
    details::check_number_of_dimensions(B, 2, "B", "outer_product_scatter_add");
    details::check_number_of_dimensions(
        indices_output, 1, "indices_output", "outer_product_scatter_add"
    );
    // Shape consistency checks are performed inside
    // mops::outer_product_scatter_add

    torch::Tensor output;
    if (A.device().is_cpu()) {
        output = torch::empty(
            {output_size, A.size(1), B.size(1)},
            torch::TensorOptions().dtype(A.scalar_type()).device(A.device())
        );

        assert(output.is_contiguous());

        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "outer_product_scatter_add", [&]() {
            mops::outer_product_scatter_add<scalar_t>(
                details::torch_to_mops_3d<scalar_t>(output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else if (A.device().is_cuda()) {
#ifndef MOPS_CUDA_ENABLED
        C10_THROW_ERROR(ValueError, "MOPS was not compiled with CUDA support " + A.device().str());
#else
        output = torch::empty(
            {output_size, A.size(1), B.size(1)},
            torch::TensorOptions().dtype(A.scalar_type()).device(A.device())
        );

        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "outer_product_scatter_add", [&]() {
            mops::cuda::outer_product_scatter_add<scalar_t>(
                details::torch_to_mops_3d<scalar_t>(output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });

#endif
    } else {
        C10_THROW_ERROR(
            ValueError, "outer_product_scatter_add is not implemented for device " + A.device().str()
        );
    }

    if (A.requires_grad() || B.requires_grad()) {
        ctx->save_for_backward({A, B, indices_output});
    }

    return {output};
}

std::vector<torch::Tensor> OuterProductScatterAdd::backward(
    torch::autograd::AutogradContext* ctx, std::vector<torch::Tensor> grad_outputs
) {
    auto saved_variables = ctx->get_saved_variables();
    auto A = saved_variables[0];
    auto B = saved_variables[1];
    auto indices_output = saved_variables[2];

    auto grad_output = grad_outputs[0].contiguous();

    auto results = OuterProductScatterAddBackward::apply(grad_output, A, B, indices_output);
    auto grad_A = results[0];
    auto grad_B = results[1];

    return {grad_A, grad_B, torch::Tensor(), torch::Tensor()};
}

std::vector<torch::Tensor> OuterProductScatterAddBackward::forward(
    torch::autograd::AutogradContext* ctx,
    torch::Tensor grad_output,
    torch::Tensor A,
    torch::Tensor B,
    torch::Tensor indices_output
) {
    auto grad_A = torch::Tensor();
    auto grad_B = torch::Tensor();

    if (A.device().is_cpu()) {
        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "outer_product_scatter_add_vjp", [&]() {
            auto mops_grad_A = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (A.requires_grad()) {
                grad_A = torch::empty_like(A);
                mops_grad_A = details::torch_to_mops_2d<scalar_t>(grad_A);
            }

            auto mops_grad_B = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (B.requires_grad()) {
                grad_B = torch::empty_like(B);
                mops_grad_B = details::torch_to_mops_2d<scalar_t>(grad_B);
            }

            mops::outer_product_scatter_add_vjp<scalar_t>(
                mops_grad_A,
                mops_grad_B,
                details::torch_to_mops_3d<scalar_t>(grad_output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else if (A.device().is_cuda()) {
#ifndef MOPS_CUDA_ENABLED
        C10_THROW_ERROR(ValueError, "MOPS was not compiled with CUDA support " + A.device().str());
#else
        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "outer_product_scatter_add_vjp", [&]() {
            auto mops_grad_A = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};

            if (A.requires_grad()) {
                grad_A = torch::empty_like(A);
                mops_grad_A = details::torch_to_mops_2d<scalar_t>(grad_A);
            }

            auto mops_grad_B = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (B.requires_grad()) {
                grad_B = torch::empty_like(B);
                mops_grad_B = details::torch_to_mops_2d<scalar_t>(grad_B);
            }

            mops::cuda::outer_product_scatter_add_vjp<scalar_t>(
                mops_grad_A,
                mops_grad_B,
                details::torch_to_mops_3d<scalar_t>(grad_output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
#endif
    } else {
        C10_THROW_ERROR(
            ValueError, "outer_product_scatter_add is not implemented for device " + A.device().str()
        );
    }

    if (grad_output.requires_grad() || A.requires_grad() || B.requires_grad()) {
        ctx->save_for_backward({grad_output, A, B, indices_output});
    }

    return {grad_A, grad_B};
}

std::vector<torch::Tensor> OuterProductScatterAddBackward::backward(
    torch::autograd::AutogradContext* ctx, std::vector<torch::Tensor> grad_outputs
) {
    auto saved_variables = ctx->get_saved_variables();
    auto grad_output = saved_variables[0];
    auto A = saved_variables[1];
    auto B = saved_variables[2];
    auto indices_output = saved_variables[3];
    auto grad_grad_A = grad_outputs[0];
    auto grad_grad_B = grad_outputs[1];

    auto grad_grad_output = torch::Tensor();
    auto grad_A_2 = torch::Tensor();
    auto grad_B_2 = torch::Tensor();

    if (A.device().is_cpu()) {
        AT_DISPATCH_FLOATING_TYPES(A.scalar_type(), "outer_product_scatter_add_vjp", [&]() {
            auto mops_grad_grad_output = mops::Tensor<scalar_t, 3>{nullptr, {0, 0, 0}};
            if (grad_output.requires_grad()) {
                grad_grad_output = torch::empty_like(grad_output);
                mops_grad_grad_output = details::torch_to_mops_3d<scalar_t>(grad_grad_output);
            }

            auto mops_grad_A_2 = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (A.requires_grad()) {
                grad_A_2 = torch::empty_like(A);
                mops_grad_A_2 = details::torch_to_mops_2d<scalar_t>(grad_A_2);
            }

            auto mops_grad_B_2 = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (B.requires_grad()) {
                grad_B_2 = torch::empty_like(B);
                mops_grad_B_2 = details::torch_to_mops_2d<scalar_t>(grad_B_2);
            }

            auto mops_grad_grad_A = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (grad_grad_A.defined()) {
                mops_grad_grad_A = details::torch_to_mops_2d<scalar_t>(grad_grad_A);
            }

            auto mops_grad_grad_B = mops::Tensor<scalar_t, 2>{nullptr, {0, 0}};
            if (grad_grad_B.defined()) {
                mops_grad_grad_B = details::torch_to_mops_2d<scalar_t>(grad_grad_B);
            }

            mops::outer_product_scatter_add_vjp_vjp<scalar_t>(
                mops_grad_grad_output,
                mops_grad_A_2,
                mops_grad_B_2,
                mops_grad_grad_A,
                mops_grad_grad_B,
                details::torch_to_mops_3d<scalar_t>(grad_output),
                details::torch_to_mops_2d<scalar_t>(A),
                details::torch_to_mops_2d<scalar_t>(B),
                details::torch_to_mops_1d<int32_t>(indices_output)
            );
        });
    } else if (A.device().is_cuda()) {
#ifndef MOPS_CUDA_ENABLED
        C10_THROW_ERROR(ValueError, "MOPS was not compiled with CUDA support " + A.device().str());
#else
        C10_THROW_ERROR(
            ValueError, "outer_product_scatter_add_vjp_vjp is not implemented for CUDA yet"
        );
#endif
    } else {
        C10_THROW_ERROR(
            ValueError,
            "outer_product_scatter_add_vjp_vjp is not implemented for device " + A.device().str()
        );
    }

    return {grad_grad_output, grad_A_2, grad_B_2, torch::Tensor()};
}
